File: phases/06-digital-logic-and-architecture/01-bits-bytes-two-s-complement-ieee-754/code/main.py
"""
Bits, Bytes, Two's Complement, IEEE 754
Phase 06 — Digital Logic & Computer Architecture

Toolkit: bit printing, two's complement negation, IEEE 754 analysis,
float comparison pitfalls.
"""
import struct


def print_bits(n: int, width: int = 32) -> str:
    """Print binary representation of n with the given bit width."""
    if n < 0:
        n = n & ((1 << width) - 1)  # reinterpret as unsigned
    bits = ""
    for i in range(width - 1, -1, -1):
        bits += str((n >> i) & 1)
        if i % 4 == 0 and i > 0:
            bits += " "
    return bits


def twos_complement_negate(n: int, bits: int = 32) -> int:
    """Negate a signed integer using two's complement: invert + 1."""
    mask = (1 << bits) - 1
    return (~n + 1) & mask


def to_signed(n: int, bits: int = 32) -> int:
    """Interpret an unsigned n-bit value as a signed two's complement integer."""
    if n >= (1 << (bits - 1)):
        return n - (1 << bits)
    return n


def sign_extend(value: int, from_bits: int, to_bits: int = 32) -> int:
    """Sign-extend a value from from_bits to to_bits."""
    sign_bit = 1 << (from_bits - 1)
    if value & sign_bit:
        mask = ((1 << to_bits) - 1) & ~((1 << from_bits) - 1)
        return value | mask
    return value & ((1 << from_bits) - 1)


def ieee754_analysis(f: float) -> dict:
    """Decompose a Python float into IEEE 754 double-precision fields."""
    # Pack as 64-bit double, unpack as uint64
    raw = struct.pack(">d", f)
    bits = int.from_bytes(raw, "big")

    sign = (bits >> 63) & 1
    exponent = (bits >> 52) & 0x7FF
    mantissa = bits & 0xFFFFFFFFFFFFF

    info = {
        "value": f,
        "hex": f"0x{bits:016X}",
        "bits": print_bits(bits, 64),
        "sign": sign,
        "sign_label": "negative" if sign else "positive",
        "exponent_raw": exponent,
        "exponent_bias": 1023,
        "exponent_actual": exponent - 1023 if exponent not in (0, 0x7FF) else None,
        "mantissa": mantissa,
    }

    if exponent == 0 and mantissa == 0:
        info["classification"] = f"{'-' if sign else '+'}zero"
    elif exponent == 0 and mantissa != 0:
        info["classification"] = "denormalized (subnormal)"
        val = (mantissa / (1 << 52)) * (2 ** (1 - 1023))
        if sign:
            val = -val
        info["computed_value"] = val
    elif exponent == 0x7FF and mantissa == 0:
        info["classification"] = f"{'-' if sign else '+'}infinity"
    elif exponent == 0x7FF and mantissa != 0:
        info["classification"] = "NaN"
    else:
        info["classification"] = "normalized"
        val = (1 + mantissa / (1 << 52)) * (2 ** (exponent - 1023))
        if sign:
            val = -val
        info["computed_value"] = val

    return info


def float_compare(a: float, b: float, rel_tol: float = 1e-9, abs_tol: float = 0.0) -> dict:
    """Demonstrate why exact float comparison fails and show robust alternatives."""
    exact = a == b
    abs_diff = abs(a - b)
    rel_diff = abs_diff / max(abs(a), abs(b)) if max(abs(a), abs(b)) > 0 else 0.0
    math_isclose = abs(a - b) <= max(rel_tol * max(abs(a), abs(b)), abs_tol)

    return {
        "a": a,
        "b": b,
        "exact_equal": exact,
        "abs_diff": abs_diff,
        "rel_diff": rel_diff,
        "math_isclose": math_isclose,
    }


def count_set_bits(n: int) -> int:
    """Count set bits using Kernighan's trick."""
    count = 0
    n = n & 0xFFFFFFFF  # limit to 32 bits
    while n:
        n &= n - 1
        count += 1
    return count


def rotate_left(n: int, shift: int, width: int = 32) -> int:
    """Rotate left by shift positions within width bits."""
    n &= (1 << width) - 1
    shift &= (width - 1)
    return ((n << shift) | (n >> (width - shift))) & ((1 << width) - 1)


def rotate_right(n: int, shift: int, width: int = 32) -> int:
    """Rotate right by shift positions within width bits."""
    n &= (1 << width) - 1
    shift &= (width - 1)
    return ((n >> shift) | (n << (width - shift))) & ((1 << width) - 1)


def safe_add(a: int, b: int, bits: int = 32) -> tuple[bool, int]:
    """Detect signed overflow using only unsigned arithmetic."""
    mask = (1 << bits) - 1
    half = 1 << (bits - 1)

    ua = a & mask
    ub = b & mask
    result = (ua + ub) & mask

    # Check if result sign disagrees with input signs
    a_sign = ua >> (bits - 1)
    b_sign = ub >> (bits - 1)
    r_sign = result >> (bits - 1)

    overflow = (a_sign == b_sign) and (a_sign != r_sign)
    signed_result = to_signed(result, bits)
    return (not overflow, signed_result)


def main() -> None:
    print("=== Two's Complement ===\n")
    for v in [0, 1, -1, 127, -128, 2147483647, -2147483648]:
        unsigned = v & 0xFFFFFFFF
        neg = twos_complement_negate(v)
        print(f"{v:>12d} : {print_bits(v)}  neg = {to_signed(neg):d}")

    print("\n=== Overflow Detection ===\n")
    ok, r = safe_add(2147483647, 1)
    print(f"INT32_MAX + 1  overflow? {'no' if ok else 'YES'}")
    ok, r = safe_add(-2147483648, -1)
    print(f"INT32_MIN + -1 overflow? {'no' if ok else 'YES'}")
    ok, r = safe_add(100, 200)
    print(f"100 + 200      overflow? {'no' if ok else 'YES'} (result = {r})")

    print("\n=== Sign Extension ===\n")
    small = to_signed(0b11111011, 8)  # -5 in 8 bits
    extended = sign_extend(0b11111011, 8, 16)
    print(f"8-bit  -5 : {print_bits(0b11111011, 8)}")
    print(f"16-bit -5 : {print_bits(extended, 16)}")

    print("\n=== IEEE 754 Double Precision ===\n")
    for val in [3.14, -6.75, 0.0, 1.0, float("inf"), float("nan")]:
        info = ieee754_analysis(val)
        print(f"value        = {info['value']}")
        print(f"hex          = {info['hex']}")
        print(f"bits         = {info['bits']}")
        print(f"sign         = {info['sign']} ({info['sign_label']})")
        print(f"exponent raw = {info['exponent_raw']}, bias = {info['exponent_bias']}, actual = {info['exponent_actual']}")
        print(f"mantissa     = 0x{info['mantissa']:013X}")
        print(f"class        = {info['classification']}")
        if "computed_value" in info:
            print(f"computed      = {info['computed_value']}")
        print()

    print("=== Bit Counting ===\n")
    for val, name in [(0, "0"), (0xFF, "0xFF"), (0xDEADBEEF, "0xDEADBEEF"), (0x80000000, "0x80000000")]:
        print(f"{name} has {count_set_bits(val)} bits set")

    print("\n=== Bitwise Operations ===\n")
    a = 0xF0F0F0F0
    b = 0x0FF00FF0
    print(f"a         = {print_bits(a)}")
    print(f"b         = {print_bits(b)}")
    print(f"a & b     = {print_bits(a & b)}")
    print(f"a | b     = {print_bits(a | b)}")
    print(f"a ^ b     = {print_bits(a ^ b)}")
    print(f"~a        = {print_bits(~a)}")
    print(f"a << 4    = {print_bits(a << 4)}")
    print(f"a >> 4    = {print_bits(a >> 4)}")

    print("\n=== Rotation ===\n")
    rot = 0x80000001
    print(f"original    = {print_bits(rot)}")
    print(f"rotl(4)     = {print_bits(rotate_left(rot, 4))}")
    print(f"rotr(4)     = {print_bits(rotate_right(rot, 4))}")

    print("\n=== Bit Tricks ===\n")
    x = 0b10110100
    print(f"x            = {print_bits(x, 8)}")
    print(f"x & (x-1)    = {print_bits(x & (x - 1), 8)}  (clear lowest set bit)")
    print(f"x & -x       = {print_bits(x & (-x), 8)}  (isolate lowest set bit)")

    print("\n=== Float Equality Pitfall ===\n")
    fa = 0.1 + 0.2
    fb = 0.3
    print(f"0.1 + 0.2 = {fa:.20f}")
    print(f"0.3        = {fb:.20f}")
    print(f"exact == ? {fa == fb}")

    comp = float_compare(0.1 + 0.2, 0.3)
    print(f"abs_diff   = {comp['abs_diff']:.2e}")
    print(f"math.isclose? {comp['math_isclose']}")

    print("\n=== Single-precision IEEE 754 (via struct) ===\n")
    raw_pi = struct.pack(">f", 3.14)
    bits_pi = int.from_bytes(raw_pi, "big")
    print(f"float 3.14f hex  = 0x{bits_pi:08X}")
    print(f"float 3.14f bits = {print_bits(bits_pi)}")
    s = (bits_pi >> 31) & 1
    e = (bits_pi >> 23) & 0xFF
    m = bits_pi & 0x7FFFFF
    print(f"sign={s} exponent={e} (actual={e-127}) mantissa=0x{m:06X}")
    print(f"reconstructed = {(1 + m / (1 << 23)) * (2 ** (e - 127)):.10f}")


if __name__ == "__main__":
    main()

File: phases/06-digital-logic-and-architecture/01-bits-bytes-two-s-complement-ieee-754/code/main.c
/*
 * Bits, Bytes, Two's Complement, IEEE 754
 * Phase 06 — Digital Logic & Computer Architecture
 *
 * Toolkit: bit printing, two's complement negation, IEEE 754
 * decomposition/construction, bit counting, rotation.
 */
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>

/* ------------------------------------------------------------------ */
/* print_bits — print binary representation of a 32-bit value         */
/* ------------------------------------------------------------------ */
void print_bits(uint32_t x) {
    for (int i = 31; i >= 0; i--) {
        printf("%d", (x >> i) & 1);
        if (i % 4 == 0 && i > 0)
            printf(" ");
    }
}

/* ------------------------------------------------------------------ */
/* negate_twos_complement — negate via invert + 1 (two's complement)  */
/* ------------------------------------------------------------------ */
int32_t negate_twos_complement(int32_t x) {
    return (int32_t)(~(uint32_t)x + 1);
}

/* ------------------------------------------------------------------ */
/* float_to_ieee754 — decompose a float into sign/exponent/mantissa   */
/* ------------------------------------------------------------------ */
void float_to_ieee754(float f) {
    uint32_t bits;
    memcpy(&bits, &f, sizeof(bits));

    uint32_t sign     = (bits >> 31) & 1;
    uint32_t exponent = (bits >> 23) & 0xFF;
    uint32_t mantissa = bits & 0x7FFFFF;

    printf("float = %.10g\n", f);
    printf("hex   = 0x%08X\n", bits);
    printf("bits  = ");
    print_bits(bits);
    printf("\n\n");

    printf("sign     = %u (%s)\n", sign, sign ? "negative" : "positive");
    printf("exponent = %u (raw), bias = 127, actual = %d\n",
           exponent, (int)exponent - 127);
    printf("mantissa = 0x%06X (fraction bits)\n\n", mantissa);

    if (exponent == 0 && mantissa == 0) {
        printf("classification: %czero\n", sign ? '-' : '+');
    } else if (exponent == 0 && mantissa != 0) {
        printf("classification: denormalized (subnormal)\n");
        /* value = (-1)^s * 0.mantissa * 2^(1-127) */
        double val = ldexp((double)mantissa / 0x800000, -126);
        if (sign) val = -val;
        printf("value         = %.10g\n", val);
    } else if (exponent == 0xFF && mantissa == 0) {
        printf("classification: %cinfinity\n", sign ? '-' : '+');
    } else if (exponent == 0xFF && mantissa != 0) {
        printf("classification: NaN (Not a Number)\n");
    } else {
        printf("classification: normalized\n");
        double val = ldexp(1.0 + (double)mantissa / 0x800000, (int)exponent - 127);
        if (sign) val = -val;
        printf("value         = %.10g\n", val);
    }
}

/* ------------------------------------------------------------------ */
/* ieee754_to_float — reconstruct a float from raw 32-bit pattern      */
/* ------------------------------------------------------------------ */
float ieee754_to_float(uint32_t bits) {
    float f;
    memcpy(&f, &bits, sizeof(f));
    return f;
}

/* ------------------------------------------------------------------ */
/* count_set_bits — Kernighan's trick: O(set bits)                    */
/* ------------------------------------------------------------------ */
uint32_t count_set_bits(uint32_t x) {
    uint32_t count = 0;
    while (x) {
        x &= x - 1;   /* clears lowest set bit */
        count++;
    }
    return count;
}

/* ------------------------------------------------------------------ */
/* rotate_left / rotate_right — circular shifts                       */
/* ------------------------------------------------------------------ */
uint32_t rotate_left(uint32_t x, unsigned n) {
    n &= 31;
    return (x << n) | (x >> (32 - n));
}

uint32_t rotate_right(uint32_t x, unsigned n) {
    n &= 31;
    return (x >> n) | (x << (32 - n));
}

/* ------------------------------------------------------------------ */
/* safe_add — detect signed overflow without UB                        */
/* ------------------------------------------------------------------ */
bool safe_add(int32_t a, int32_t b, int32_t *result) {
    uint32_t ua = (uint32_t)a;
    uint32_t ub = (uint32_t)b;
    uint32_t sum = ua + ub;

    /* Overflow when both operands have the same sign but result differs. */
    /* For positive overflow: a > 0, b > 0, result < 0 */
    /* For negative overflow: a < 0, b < 0, result >= 0 */
    if ((a > 0 && b > 0 && (int32_t)sum < 0) ||
        (a < 0 && b < 0 && (int32_t)sum >= 0)) {
        return false;  /* overflow */
    }
    *result = (int32_t)sum;
    return true;
}

/* ------------------------------------------------------------------ */
/* main — demonstrations                                              */
/* ------------------------------------------------------------------ */
int main(void) {
    printf("=== Two's Complement ===\n\n");

    int32_t vals[] = {0, 1, -1, 127, -128, INT32_MAX, INT32_MIN};
    int n = sizeof(vals) / sizeof(vals[0]);
    for (int i = 0; i < n; i++) {
        int32_t v = vals[i];
        printf("%12d : ", v);
        print_bits((uint32_t)v);
        printf("  neg = %d\n", negate_twos_complement(v));
    }

    printf("\n=== Overflow Detection ===\n\n");
    int32_t result;
    printf("INT32_MAX + 1  overflow? %s\n",
           safe_add(INT32_MAX, 1, &result) ? "no" : "YES");
    printf("INT32_MIN + -1 overflow? %s\n",
           safe_add(INT32_MIN, -1, &result) ? "no" : "YES");
    printf("100 + 200      overflow? %s (result = %d)\n",
           safe_add(100, 200, &result) ? "no" : "YES", result);

    printf("\n=== Sign Extension ===\n\n");
    int8_t  small = -5;
    int16_t wide  = small;     /* sign-extended by compiler */
    printf("int8_t  -5 : "); print_bits((uint32_t)(uint8_t)small);  printf("\n");
    printf("int16_t -5 : "); print_bits((uint32_t)(uint16_t)wide);  printf("\n");

    printf("\n=== IEEE 754 Floats ===\n\n");
    float_to_ieee754(3.14f);
    printf("\n");
    float_to_ieee754(-6.75f);
    printf("\n");
    float_to_ieee754(0.0f);
    printf("\n");
    float_to_ieee754(1.0f / 0.0f);   /* +inf */
    printf("\n");
    float_to_ieee754(0.0f / 0.0f);   /* NaN */

    printf("\n=== Round-trip: bits -> float -> bits ===\n\n");
    uint32_t raw = 0x40490FDB;  /* pi approx */
    float reconstructed = ieee754_to_float(raw);
    printf("raw = 0x%08X -> float = %.10g\n", raw, reconstructed);

    printf("\n=== Bit Counting ===\n\n");
    uint32_t demo_vals[] = {0, 0xFF, 0xDEADBEEF, 0x80000000};
    const char *demo_names[] = {"0", "0xFF", "0xDEADBEEF", "0x80000000"};
    for (int i = 0; i < 4; i++) {
        printf("%s has %u bits set\n",
               demo_names[i], count_set_bits(demo_vals[i]));
    }

    printf("\n=== Bitwise Operations ===\n\n");
    uint32_t a = 0xF0F0F0F0;
    uint32_t b = 0x0FF00FF0;
    printf("a         = "); print_bits(a); printf("\n");
    printf("b         = "); print_bits(b); printf("\n");
    printf("a & b     = "); print_bits(a & b); printf("\n");
    printf("a | b     = "); print_bits(a | b); printf("\n");
    printf("a ^ b     = "); print_bits(a ^ b); printf("\n");
    printf("~a        = "); print_bits(~a); printf("\n");
    printf("a << 4    = "); print_bits(a << 4); printf("\n");
    printf("a >> 4 (u)= "); print_bits(a >> 4); printf("\n");

    printf("\n=== Rotation ===\n\n");
    uint32_t rot = 0x80000001;
    printf("original    = "); print_bits(rot); printf("\n");
    printf("rotl(4)     = "); print_bits(rotate_left(rot, 4)); printf("\n");
    printf("rotr(4)     = "); print_bits(rotate_right(rot, 4)); printf("\n");

    printf("\n=== Bit Tricks ===\n\n");
    uint32_t x = 0b10110100;
    printf("x            = "); print_bits(x); printf("\n");
    printf("x & (x-1)    = "); print_bits(x & (x - 1)); printf("  (clear lowest set bit)\n");
    printf("x & -x       = "); print_bits(x & (uint32_t)(-(int32_t)x)); printf("  (isolate lowest set bit)\n");

    printf("\n=== Float Equality Pitfall ===\n\n");
    float sum = 0.0f;
    for (int i = 0; i < 10; i++)
        sum += 0.1f;
    printf("0.1f added 10 times = %.20f\n", sum);
    printf("1.0f                 = %.20f\n", 1.0f);
    printf("equal? %s (accumulated rounding error)\n",
           sum == 1.0f ? "yes" : "NO");

    return 0;
}
